records.go

```diff
@@ -33,16 +37,55 @@ func (re *Records) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Ms
 	m.Authoritative = true
 
 	nxdomain := true
+	// cnameMaybeUpstream tracks whether we are currently trying to resolve a CNAME. We always look for a match among
+	// the records handled by this plugin first, then we go upstream. This is required to enforce stack depth and loop
+	// detection.
+	cnameMaybeUpstream := false
 	var soa dns.RR
+	cnameStack := make(map[string]struct{}, 0)
+
+resolveLoop:
 	for _, r := range re.m[zone] {
+		if _, ok := cnameStack[qname]; ok {
+			log.Errorf("detected loop in CNAME chain, name [%s] already processed", qname)
+			goto servfail
+		}
+		if len(cnameStack) > maxCnameStackDepth {
+			log.Errorf("maximum CNAME stack depth of %d exceeded", maxCnameStackDepth)
+			goto servfail
```

Comment:
Not sure if servfail is right answer here.  It could be, but I think a response with a dangling cname might be preferable to a servfail.  Not sure, need to check RFCs/other implementations.